design/multi-checkout.md

```diff
@@ -0,0 +1,180 @@
+Therefore we have this stake in the ground:
+
+> We are in "multi-checkout" when the customer has explicitly asked for multiple repositories to be checked out.
+> Zero checkout steps yields the same behavior as today (checkout `self` in a normal job, no checkout in a `deployment` job).
+> One checkout (whether it's `self` or not) preserves existing single-repo behavior.
+
+## Multi-checkout mode
+
+_(Note: I'm just calling it a "mode" for purposes of discussion here. We won't document it as a "mode".)_
+
+The agent will look at the job message and discover all `checkout` steps.
+
+### Zero or one checkout step
+0. If there are none at all, we'll preserve the existing default behavior, acting as if a `checkout: self` were the first step.
+0. If there is exactly one `checkout: none`, then we'll preserve the existing behavior: no repositories are synced or checked out.
+0. If there is exactly one `checkout` step that isn't `none`, then we'll use existing single-checkout behavior but check out that repository rather than `self`.
+
+### More than one checkout step
+0. If there are multiple checkout steps and one or more are `checkout: none`, this is an error.
+It's unclear what the user meant for us to do, so we fail with a clear message.
+0. Otherwise, we are in multi-checkout mode.
+
+### Behavior changes in multi-checkout mode
+
+#### Checkout directory
+By default, all repositories will be checked out as if they've been `git clone`d into the `Pipeline.Workspace` directory.
+Git uses the last part of the path, minus any trailing slashes, spaces, and the string `.git`.
+(We have YAML syntax for selecting a different path, and if the user specifies that, we'll use their name.)
+This changes the default directory for the `self` repo: it would be `$(Pipeline.Workspace)/s` if it were the only `checkout` step.
+
+#### System variables
+`Build.SourcesDirectory` points to the directory where the `self` repo was checked out whether in single- or multi-checkout mode.
+`System.DefaultWorkingDirectory` points to the same place as `Pipeline.Workspace` in multi-checkout mode (which is different than single-checkout mode).
+We believe this preserves the intended semantics of these variables, keeping as many scripts and tasks working as possible.
+
+## Resource specification
+
+The `resources` block already allows a `repositories` list.
+Today, those repositories hold only YAML templates, so we can't blindly check them out automatically.
+That's why an explicit `checkout` step is also required.
+
+However, in multi-checkout scenarios, this leads to unnecessary verbosity.
+First you have to mention the repository up top, give it a name, and then you check it out by name.
+To combat this, we'll add an inline repository syntax.
+Inline syntax will cover the `type`, `name`, and `ref` of a repository resource.
+(If other changes are needed - such as `path` or `endpoint` - users must fall back to explicit resource syntax.)
```

Comment:
`path` is a little ambiguous here since there is a `path` on the checkout task. Is there also one on the resource as well?

Reply:
Oops typo. Will fix.